File: src/main/java/alfheim/common/entity/boss/ai/AIBase.java
package alfheim.common.entity.boss.ai;

import alexsocol.asjlib.ASJUtilities;
import alfheim.common.entity.boss.EntityFlugel;
import net.minecraft.entity.ai.EntityAIBase;

public abstract class AIBase extends EntityAIBase {

	EntityFlugel flugel;
	AITask task;
	
	public AIBase(EntityFlugel flugel, AITask task) {
		this.flugel = flugel;
		this.task = task;
		setMutexBits(5);
	}
	
	@Override
	public boolean shouldExecute() {
		// ASJUtilities.chatLog("Should " + task.toString() + ": " + (flugel.getHealth() > 0 && flugel.getAITask().equals(task) && flugel.getAITaskTimer() == 0));
		return flugel.getHealth() > 0 && flugel.getAITask().equals(task) && flugel.getAITaskTimer() == 0;
	}

	@Override
	public abstract void startExecuting();
	
	public boolean canContinue() {
		if (flugel.getHealth() <= 0 || !flugel.getAITask().equals(task)) return false;
		flugel.setAITaskTimer(flugel.getAITaskTimer() - 1);
		return flugel.getAITaskTimer() > 0;
	}
	
	@Override
	public abstract boolean continueExecuting();
	
	@Override
	public void resetTask() {
		flugel.setAITask(AITask.getRand());
	}	
}

File: src/main/java/alfheim/common/entity/boss/ai/AIBase.kt
package alfheim.common.entity.boss.ai;

import alfheim.common.entity.boss.EntityFlugel
import net.minecraft.entity.ai.EntityAIBase

abstract class AIBase(val flugel: EntityFlugel, val task: AITask) : EntityAIBase() {
	
	init {
		setMutexBits(5)
	}

	override fun shouldExecute(): Boolean = flugel.getHealth() > 0 && flugel.getAITask().equals(task) && flugel.getAITaskTimer() == 0

	override abstract fun startExecuting()
	
	fun canContinue(): Boolean {
		if (flugel.getHealth() <= 0 || !flugel.getAITask().equals(task)) return false
		flugel.setAITaskTimer(flugel.getAITaskTimer() - 1)
		return flugel.getAITaskTimer() > 0
	}

	override abstract fun continueExecuting(): Boolean
	
	override fun resetTask() {
		flugel.setAITaskTimer(0)
		flugel.setAITask(flugel.nextTask())
	}
}